### PR TITLE
Commerce 7829 UI

### PR DESCRIPTION
**Pull request by andrea-ale-sbarra**
[JIRA Ticket](https://issues.liferay.com/browse/COMMERCE-7829)

### Context

To discontinue a product means just mark the sku as discontinued. The scope of this new feature is to let the buyer to purchase the product regardless of the item that is actually added to cart. In this use case scenario the admin user will configure the product so that it will bemarked as discontinued and if a replacement item is configured, the buyer can still purchase the product via its replacement.

### CODE REVIEW

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java [hunk 1]
```diff
@@ -957,6 +981,28 @@ public CPInstance updateCPInstance(
 		CPInstance cpInstance = cpInstancePersistence.findByPrimaryKey(
 			cpInstanceId);
 
+		int discontinuedDateMonth = 0;
+		int discontinuedDateDay = 0;
+		int discontinuedDateYear = 0;
+		int discontinuedDateHour = 0;
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPInstanceLocalServiceImpl.java]: remove hour and minutes

**Comment by andrea-ale-sbarra** [comment]
ci:test:sf

**Comment by andrea-ale-sbarra** [comment]
ci:test:sf

**Comment by andrea-ale-sbarra** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java [hunk 2]
```diff
@@ -177,6 +177,25 @@ public long getCPInstanceReplacementId() throws PortalException {
 		return cpInstanceReplacement.getCPInstanceId();
 	}
 
+	public String getCPInstanceReplacementLabel() throws PortalException {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java:180]: getReplacementCPInstanceSku()

File: modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java [hunk 3]
```diff
@@ -177,6 +177,25 @@ public long getCPInstanceReplacementId() throws PortalException {
...
+		CPInstance cpInstance = getCPInstance();
+
+		if (cpInstance == null) {
+			return StringPool.BLANK;
+		}
+
+		CPInstance cpInstanceReplacement =
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java:187]: replacementCPInstance instead

File: modules/apps/commerce/commerce-product-definitions-web/src/main/resources/META-INF/resources/instance/details.jsp [hunk 4]
```diff
@@ -339,24 +371,39 @@ if ((cpInstance != null) && (cpInstance.getExpirationDate() != null)) {
...
-		},
+	});
+
+	const discontinuedInput = document.getElementById(
+		'<%= liferayPortletResponse.getNamespace() %>discontinued'
+	);
+	const discontinuedDateInput = document.getElementById(
+		'<%= liferayPortletResponse.getNamespace() %>discontinuedDate'
+	);
+	const replacementIdWrapper = document.getElementById(
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/resources/META-INF/resources/instance/details.jsp:393]: Call it replacementAutocompleteWrapper instead

File: modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java [hunk 5]
```diff
@@ -158,6 +158,24 @@ public long getCPInstanceId() throws PortalException {
 		return cpInstance.getCPInstanceId();
 	}
 
+	public long getCPInstanceReplacementId() throws PortalException {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java:161]: getReplacementCPinstanceId() instead

File: modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java [hunk 6]
```diff
@@ -158,6 +158,24 @@ public long getCPInstanceId() throws PortalException {
...
+		CPInstance cpInstance = getCPInstance();
+
+		if (cpInstance == null) {
+			return 0;
+		}
+
+		CPInstance cpInstanceReplacement = _cpInstanceHelper.fetchCPInstanceReplacement(cpInstance.getReplacementCProductId(), cpInstance.getReplacementCPInstanceUuid());
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/display/context/CPInstanceDisplayContext.java]: replacementCPInstance

File: modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/portlet/action/EditCPInstanceMVCActionCommand.java [hunk 7]
```diff
@@ -263,11 +263,11 @@ protected CPInstance updateCPInstance(ActionRequest actionRequest)
 		String replacementCPInstanceUuid = null;
 		long replacementCProductId = 0;
 
-		long skuId = ParamUtil.getLong(actionRequest, "skuId");
+		long cpInstanceReplacementId = ParamUtil.getLong(actionRequest, "cpInstanceReplacementId");
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/java/com/liferay/commerce/product/definitions/web/internal/portlet/action/EditCPInstanceMVCActionCommand.java]: Call it replacementCPInstanceId

File: modules/apps/commerce/commerce-product-definitions-web/src/main/resources/META-INF/resources/instance/details.jsp [hunk 8]
```diff
@@ -22,6 +22,7 @@ CPInstanceDisplayContext cpInstanceDisplayContext = (CPInstanceDisplayContext)re
 CPDefinition cpDefinition = cpInstanceDisplayContext.getCPDefinition();
 CPInstance cpInstance = cpInstanceDisplayContext.getCPInstance();
 long cpInstanceId = cpInstanceDisplayContext.getCPInstanceId();
+long cpInstanceReplacementId = cpInstanceDisplayContext.getCPInstanceReplacementId();
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-definitions-web/src/main/resources/META-INF/resources/instance/details.jsp]: replacementCPInstanceId

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/util/CPInstanceHelperImpl.java [hunk 9]
```diff
@@ -99,6 +99,15 @@ public CPInstance fetchCPInstance(
 			cpDefinitionId, serializedDDMFormValues);
 	}
 
+	@Override
+	public CPInstance fetchCPInstanceReplacement(
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/util/CPInstanceHelperImpl.java:103]: fetchReplacementCPInstance(long cProductId, String cpInstanceUuid) instead

File: modules/apps/commerce/commerce-product-api/src/main/java/com/liferay/commerce/product/util/CPInstanceHelper.java [hunk 10]
```diff
@@ -43,6 +43,10 @@ public CPInstance fetchCPInstance(
 			long cpDefinitionId, String serializedDDMFormValues)
 		throws PortalException;
 
+	public CPInstance fetchCPInstanceReplacement(
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-api/src/main/java/com/liferay/commerce/product/util/CPInstanceHelper.java:46]: fetchReplacementCPInstance(long cProductId, String cpInstanceUuid) instead

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@andrea-ale-sbarra In general we want to prefix CPInstance with "replacement" instead
Thanks

**Comment by andrea-ale-sbarra** [comment]
ci:forward

**Comment by andrea-ale-sbarra** [comment]
ci:forward

**Comment by andrea-ale-sbarra** [comment]
ci:test:commerce-functional

**Comment by andrea-ale-sbarra** [comment]
ci:forward

**Pull request closed without merging**